### PR TITLE
`GradScaler` recomputes `optimizer_state["found_inf_per_device"]` before `optimizer.step`

### LINKED ISSUE #96755: Fused AdamW has worse loss than Apex and unfused AdamW for fp16/AMP
### 🐛 Describe the bug

@stas00 recently benchmarked on an 80GB A100 and discovered fused AdamW is worse than both the Apex equivalent and the default implementation. Moreover, fused AdamW is not more performant in this case.
 
| Variation                 |     Train samples per second |   Diff % |   Train loss |
|:--------------------------|----------:|-------:|--------:|
| --optim adamw_torch_fused |    389.41 |      0 |    2.66 |
| --optim adamw_torch       |    389.37 |      0 |    2.55 |
| --optim adamw_apex_fused  |    399.27 |      3 |    2.53 |

Repro:
See @stas00's comment below!
The batch size (--per_device_train_batch_size) can be adjusted to be smaller.

cc @ezyang @gchanan @zou3519 @vincentqb @jbschlosser @albanD @crcrpar

### Versions

Master/latest nightly

### PR DESCRIPTION
**Pull request by crcrpar**
I found a discrepancy between non-fused and fused optimizers, which is to use `optimizer_state["found_inf"]` or to recompute `found_inf`.

- non fused: https://github.com/pytorch/pytorch/blob/e64ddd1ab9d46cfc921c19269969ffc5cd7d6f6c/torch/cuda/amp/grad_scaler.py#L289
- fused: https://github.com/pytorch/pytorch/blob/e64ddd1ab9d46cfc921c19269969ffc5cd7d6f6c/torch/cuda/amp/grad_scaler.py#L353
    - where `_check_inf_per_device` is https://github.com/pytorch/pytorch/blob/e64ddd1ab9d46cfc921c19269969ffc5cd7d6f6c/torch/cuda/amp/grad_scaler.py#L564-L573

The other way to align the behavior is to use the existing `found_inf` in https://github.com/pytorch/pytorch/blob/e64ddd1ab9d46cfc921c19269969ffc5cd7d6f6c/torch/cuda/amp/grad_scaler.py#L353.

I'd say this PR is for the sake of "safety" and the alternative is to keep the existing behavior.
I honestly have no idea if it's expected to double-check the sanity of gradients in `GradScaler.step`.

---

what I've observed in huggingface/transformers T5-base example so far seems like that non-fused optimizers lead to invalid parameters while the fused not. 
The cause seems to be that `gradients` become inf/nan before `GradScaler.step(optimizer)` after `GradScaler._unscale_grads_` (more precicely, the call of `torch._amp_foreach_non_finite_check_and_unscale_`) in the script of the issue linked below, i.e. the gradient clipping and/or unscaling lead to inf/nan as these happen after the grad check. See
https://github.com/pytorch/pytorch/blob/788300cc2aa096d8d5c1e7fbfc87e5439a338251/aten/src/ATen/native/cuda/AmpKernels.cu#L165-L174.

Fixes #96755 🙏 

cc @vincentqb @jbschlosser @albanD @janeyx99 @mcarilli @ptrblck @leslie-fang-intel @jgong5

### CODE REVIEW

**Comment by crcrpar** [comment]
looks okay-ish. will add a test case

*** Results:

| Variation                 |   Train<br>samples<br>per<br>second |   Diff<br>% |   Train<br>loss |
|:--------------------------|------------------------------------:|------------:|----------------:|
| --optim adamw_torch_fused |                              332.32 |           2 |            2.87 |
| --optim adamw_torch       |                              326.27 |           0 |            2.87 |
| --optim adamw_apex_fused  |                              337.26 |           3 |            2.87 |

```

*** Setup:


Datetime    : 2023-03-23 01:50:40

Software:
transformers: 4.28.0.dev0
torch       : 2.1.0a0+git517a432
cuda        : 12.1
python      : 3.10.9

Hardware:
1 GPUs      : NVIDIA A100-SXM-80GB, 79.35GB


*** The benchmark command line was:

CUDA_VISIBLE_DEVICES=0 python scripts/benchmark/trainer-benchmark.py \
--base-cmd \
' \
examples/pytorch/translation/run_translation.py --model_name_or_path t5-base --output_dir output_dir \
--do_train --label_smoothing 0.1 --logging_strategy no --save_strategy no --per_device_train_batch_size 32 \
--max_source_length 512 --max_target_length 512 --num_train_epochs 1 --overwrite_output_dir \
--source_lang en --target_lang ro --dataset_name wmt16 --dataset_config "ro-en" \
--source_prefix "translate English to Romanian: "  --warmup_steps 50 \
--max_train_samples 30000 --dataloader_num_workers 2 --fp16 \
' \
--target-metric-key train_samples_per_second --repeat-times 1 --variations \
'--optim adamw_torch_fused|--optim adamw_torch|--optim adamw_apex_fused' \
--report-metric-keys train_loss --base-variation '--optim adamw_torch'

```

**Review by ngimel — APPROVED** [review]
Thanks for the fix!

**Comment by ngimel** [comment]
cc @janeyx99

**Review by janeyx99 — APPROVED** [review]

File: test/test_cuda.py [hunk 1]
```diff
@@ -2525,6 +2525,43 @@ def _grad_scaling_autocast_fused_optimizers(self, optimizer_ctor, optimizer_kwar
                         actual = actual.squeeze()
                     self.assertEqual(state_control[k], actual)
 
+    def test_grads_invalidated_between_unscale_and_step(self):
```
[the remark this answers is not in the document]
Comment (janeyx99) [reply test/test_cuda.py:2528]: could also be done in another PR, not blocking

[the remark this answers is not in the document]
Comment (crcrpar) [reply test/test_cuda.py:2528]: I don't think so because https://github.com/pytorch/pytorch/blob/5f5d675587877142801c50e7dc91e41df1ea9eaa/test/test_cuda.py#L47-L52 looks sufficient

**Comment by janeyx99** [comment]
@crcrpar Could you cherry-pick this PR following the instructions in https://github.com/pytorch/pytorch/issues/97272 for the potential 2.0.1 release? The category here I believe is 2.

**Comment by ngimel** [comment]
Hey @crcrpar, we have some multi-gpu jobs getting stuck after this diff. Does the change add some multi-gpu communications or changes where they are happening?

**Comment by janeyx99** [comment]
@crcrpar I do have a question that may or may not be relevant:

It looks like there is a discrepancy in the path where the optim handles grad_scale and found_inf vs the path where the optimizer does not:

| **Path where optim handles grad_scale and found_inf (Fused)**          | **Path where optim does not (Unfused)**                                                |
|------------------------------------------------------------------------|----------------------------------------------------------------------------------------|
| _check_scale_growth_tracker                                            | _check_scale_growth_tracker                                                            |
| _check_inf_per_device (calls unscale with 1, so harmless)              | unscale_ WHICH ACTUALLY UNSCALES WITH SCALE RECIPROCAL REGARDLESS OF WHAT FOUND_INF IS |
| optimizer.step() WHICH ONLY UNSCALES THE GRADS WHEN FOUND_INF is not 1 | _maybe_opt_step                          |

This looks like the fused implementation will _miss out on unscaling the grads_ when found_inf is 1 whereas the other implementations will _unscale first_ and then skip the step. This shouldn't matter if zero_grad is called before every iteration, but could this possibly cause inconsistencies?

**Comment by crcrpar** [comment]
Excuse me but I don't think I understand your point.

> This shouldn't matter if zero_grad is called before every iteration, but could this possibly cause inconsistencies?

What did you mean by "every iteration"?
More specifically, iteration here sounds a bit unclear to me.
Did you mean a period between each parameter updates? Would you assume, for example, gradient accumulation?
Would you assume a situation where we call [`GradScaler.step`](https://pytorch.org/docs/stable/amp.html#torch.cuda.amp.GradScaler.step) w/o zeroing out gradients the call?

**Comment by janeyx99** [comment]
My point was that I observed an inconsistency that _shouldn’t_ matter, but I wanted to get your thoughts on it. By iteration here I mean a “forward, backward, optimizer step”. If zero_grad is called after the step or before the forward, I do not think polluting the grads will matter. Basically I’m curious if I can this is inconsistency can affect more than I believe it could.

**Comment by crcrpar** [comment]
I still am not clear. If gradients are not flushed after parameter update, I'd think few training recipes work.

**Comment by janeyx99** [comment]
@crcrpar Yea, that's what my assumption is too. I wanted to see if I might have been missing something, but your responses confirm this is likely not relevant then.

**Comment by ngimel** [comment]
The docs have a section on using GradScaler with gradient accumulation https://pytorch.org/docs/stable/notes/amp_examples.html#gradient-accumulation, which states that `zero_grad` should be called after every parameter update. If there are infs in gradients, step will be skipped, so it doesn't matter whether unscale was applied, so I think the difference in behavior that @janeyx99 described won't matter (though it is indeed a difference!).
